### ISSUE TITLE
Cross Encoder BCE Loss Activation

### ISSUE THREAD

**Issue by OsamaS99** [UKPLab/sentence-transformers#3326]
Hello Tom,

So I'm currently training a cross-encoder for re-ranking task. So I just wanted to ask about couple things:

- Does the output of a cross encoder pass first through a sigmoid-function when wrapped with CrossEncoder?
- I noticed that for BCE loss, the default activation is `nn.Identity()`, that isn't correct as far as I know. See [pytorch](https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html). BCE is supposed to decrease the NLL (Negative Log Likelihood) of a Bernoulli distribution, therefore the input to the loss should be probabilities.
- Do you think it should default to other losses as well?

Best,
Osama

**Comment by tomaarsen** [comment]
Hello!

> * Does the output of a cross encoder pass first through a sigmoid-function when wrapped with CrossEncoder?

Not during training, no. During inference, the `predict` method uses the `activation_fn` option to decide what activation function is used. It defaults to Sigmoid, but you can use e.g. Identity and then save the model (it's stored in the `config.json` so your option is saved).
During training, the `predict` method isn't used, and the only activation function that's used is the one in the loss function.

> * I noticed that for BCE loss, the default activation is `nn.Identity()`, that isn't correct as far as I know. See [pytorch](https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html). BCE is supposed to decrease the NLL (Negative Log Likelihood) of a Bernoulli distribution, therefore the input to the loss should be probabilities.

I'm using the [`BCEWithLogitsLoss`](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html) instead of the `BCELoss`, see e.g.: https://stackoverflow.com/questions/75979632/pytorchs-nn-bcewithlogitsloss-behaves-totaly-differently-than-nn-bceloss

Notably: "This loss combines a Sigmoid layer and the BCELoss in one single class. This version is more numerically stable than using a plain Sigmoid followed by a BCELoss as, by combining the operations into one layer, we take advantage of the log-sum-exp trick for numerical stability."

> * Do you think it should default to other losses as well?

I don't think I'm following. Do you mean that it should default to Sigmoid as the activation function for all losses? We use that activation function for e.g. MultipleNegativesRankingLoss and CachedMultipleNegativesRankingLoss. For example the [mGTE paper](https://arxiv.org/pdf/2407.19669) uses this loss (called InfoNCE with Cached Gradients there) but with the Tanh activation function.


- Tom Aarsen

**Comment by OsamaS99** [comment]
Hello Tom,

Thank you very much for clarification, that answers my questions.
